common/src/main/java/org/opensearch/sdk/client/LocalClusterIndicesClient.java

```diff
@@ -65,6 +70,9 @@
 @Log4j2
 public class LocalClusterIndicesClient implements SdkClientDelegate {
 
+    private static final String HASH_KEY = "_tenant_id";
```

Comment:
Rename to `tenant_id`? By default there would be no _tenant_id field

Reply:
I literally copied and pasted this line of code from the DDBClient.  Granted, that was changed in a recent PR because I didn't realize it was defined elsewhere: we should not be duplicating important strings like this in code, this is a classic example of why; I missed that we have a `CommonValue` for it and would have changed it instead.

I still maintain `_tenant_id` is better and plan to change the `CommonValue` string to align.  Why do prefer a plain field that any regular user can accidentally overwrite with a sensible sounding field?  Are we taking any action to prevent users from not using these special field names?

We use leading underscores for other important metadata like document Id (`_id`), as well as `_seq_no` and `_primary_term` and `_source`.   It's a well known pattern.  What if the user wants to actually use their own "tenant ID" in a document? How will they know that using `tenant_id` will end up denying them access to their documents?  Are there any other secret keys that will totally break their documents if they use them?

Let's align on this and

Reply:
> We use leading underscores for other important metadata like document Id (_id), as well as _seq_no and _primary_term and _source. It's a well known pattern. What if the user wants to actually use their own "tenant ID" in a document? 

This makes sense, we should mark the metadata fields in an document in an `internal` approach with leading underscore.

Reply:
The reason I put `tenant_id` as a index field because other fields in the ml-commons system indices don't have leading `_`.  So only putting `_` for the `tenant_id` does break the consistency with other fields.

Now I feel like `_id`, `_seq_no`, `_primary_term`, `_primary_term` all these are coming from core. What should be the rationale of adding `_` when plugin is introducing this field?

>What if the user wants to actually use their own "tenant ID" in a document?

This problem can happen for any fields we defined in the ml-commons indices, right? Not just for `tenant_id`

>we should mark the metadata fields in an document in an internal approach with leading underscore.

what does `internal` mean here? Internal to whom? Plugin or Core?

Reply:
> > we should mark the metadata fields in an document in an internal approach with leading underscore.
> 
> what does `internal` mean here? Internal to whom? Plugin or Core?

A leading underscore in a variable name is a long understood coding convention for internal variables.  
 - Python: https://realpython.com/python-double-underscore/
 - C / C++: https://devblogs.microsoft.com/oldnewthing/20230109-00/?p=107685

It's not enforced, it's not mandatory, but it's what people expect.  Style guides say to not use those types of variable names -- which is precisely why you use them when you don't want to conflict with existing user present or future variable names.

Tell me: of all OpenSearch users now and ever in the future, how will you guarantee that no user will ever use the field name `tenant_id` in a document.

I'd wager many already have; and many will in the future.

> The reason I put `tenant_id` as a index field because other fields in the ml-commons system indices don't have leading `_`. So only putting `_` for the `tenant_id` does break the consistency with other fields.

If it's ml plugin specific then give it a ml plugin specific name.  There just needs to be a common consistent way to tell users "hey, don't use this field name or you'll break things".   Leading underscores is already a well understood way.  If you want to do something else, have a common prefix that you tell users to avoid.    Particularly if you use a non-verbose, very common identifier such as "tenant ID". 

> This problem can happen for any fields we defined in the ml-commons indices, right? Not just for `tenant_id`

Yes.   How are you going to tell users not to overwrite it?  Options are:
 - tell users not to start fields with `_`. (That's already an expectation.)
 - tell users not to start fields with `ml_plugin_`.  Seems reasonable.
 - maintain a list somewhere of reserved names, and include all future names that we'll ever want to use so that we don't break existing indices that might already have those names.   Seems impossible.

Reply:
If this is metadata in system index, then we don't need to force the leading underscore, since they'll not stored in user's index(not mixed with user data). For `_id`, `_seq_no `, they're mixed with user data, so better use `_` as persevered fields.

Reply:
> If this is metadata in system index, then we don't need to force the leading underscore, since they'll not stored in user's index(not mixed with user data). 

For this particular implementation we are using system indices only in ML plugins.

There are other plugins who may want to use a generic term "tenant_id" for a completely different purpose.  The name should either be more specific, or configurable by a plugin.

Reply:
@zane-neo I looked into wrapping the `actionGet()` in another async thread. Here's what it would look like:
```
    @Override
    public CompletionStage<SearchDataObjectResponse> searchDataObjectAsync(
        SearchDataObjectRequest request,
        Executor executor,
        Boolean isMultiTenancyEnabled
    ) {
        log.info("Searching {}", Arrays.toString(request.indices()));
        SearchSourceBuilder searchSource = request.searchSourceBuilder();
        if (Boolean.TRUE.equals(isMultiTenancyEnabled)) {
            QueryBuilder existingQuery = searchSource.query();
            String tenantId = request.tenantId() != null ? request.tenantId() : CommonValue.DEFAULT_TENANT;
            BoolQueryBuilder boolQuery = QueryBuilders.boolQuery().must(existingQuery == null ? new MatchAllQueryBuilder() : existingQuery);
            boolQuery.filter(QueryBuilders.termQuery(CommonValue.TENANT_ID, tenantId));
            searchSource.query(boolQuery);
        }
        log.info("Searching {}", Arrays.toString(request.indices()));
        ActionFuture<SearchResponse> searchResponseFuture = AccessController
            .doPrivileged((PrivilegedAction<ActionFuture<SearchResponse>>) () -> {
                return client.search(new SearchRequest(request.indices(), searchSource));
            });
        return CompletableFuture
            .supplyAsync(searchResponseFuture::actionGet, executor)
            .thenApply(searchResponse -> {
                log.info("Search returned {} hits", searchResponse.getHits().getTotalHits());
                try {
                    return SearchDataObjectResponse.builder().parser(createParser(searchResponse)).build();
                } catch (IOException e) {
                    // Rethrow unchecked exception on XContent parsing error
                    throw new OpenSearchStatusException(
                        "Failed to search indices " + Arrays.toString(request.indices()),
                        RestStatus.INTERNAL_SERVER_ERROR
                    );
                }
            });
    }
```
I'm not sure it really adds anything here.

At the moment I'm still leaning toward leaving the code as is.  Let me know if you have additional concerns.

Reply:
This is not async because this line still block waiting: `.supplyAsync(searchResponseFuture::actionGet, executor)`. But if user can somehow set the timeout to this client, then it's fine because it won't exhausted the threads in thread pool.

Reply:
If we don't rely on good intention, which means if user doesn't set the timeout to the client, what happens?

Reply:
The same thing that happens on OpenSearch NodeClient search request with no timeout.

The executor thread keeps waiting.

In OpenSearch it's the ActionListener thread.  Here it's a user-settable executor.